### PR TITLE
Missing advertisement label

### PR DESCRIPTION
**Pull request by janua**
## What does this change?

When we either load Programmatic guaranteed or Native backfill, these advertisements come without an advertisement label at the top of the advert as they have their own language to communicate they are an advert.

Once one of these gets displayed we eventually then refresh the advert at a later point. This causes the label to disappear forever for the current pageview.

The main detail around this is that `ad-slot--gc` gets added to the ad-slot when one of these native ad types get displayed, and subsequent loads/refreshes of the ad-slot does not remove the added class.

#### Technical Details

There are two main problems:
 - When we render an ad-slot, we add classes to the ad-slot element that give it appropriate styling. With ad-refresh these classes never get re-evaluated as to whether they are appropriate for the ad-slot or not and instead hang onto the ad-slot forever. We need to remove the classes during an ad-refresh that are no longer appropriate for the next ad impression. This is fixed by tracking which classes come in and out of the `Advert` class. 
 - `render-advert-label.renderAdvertLabel` does not use `fastdom` to synchronise reading/writing of the DOM, nor does it wait for any `fastdom` promises to finish before doing it's check. This means when we call `fastdom` in `render-advert.sizeCallbacks[...]`, by the time we call `renderAdvertLabel` the DOM has not updated yet. In some cases the advertisement label always gets inserted into the DOM as the check to prevent this from happening will always pass.
 

#### Bigger issue (ish)
There is a bigger issue; we use both Javascript **AND** CSS to hide the advertisement label. With Javascript this happens with the check in `renderAdvertLabel`explained above always passing meaning it always displayed.
This almost makes the `render-advert-label` module redundant (apart from the feedback part of it), as both CSS and JS are doing the same thing in a different way. We should consider using CSS for hiding the label in circumstances where it should not appear instead of depending on Javascript to not insert it.

Here is another way of understanding what I am saying;

 - 1. We check in Javascript [here](https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js#L12) if we should insert the advertisement label.
 - 2. Because this is neither in `fastdom` or chained from the [promises that write the appropriate class to the DOM](https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/commercial/modules/dfp/render-advert.js#L108), it always passes, meaning it always gets inserted into the ad-slot.
 - 3. What the Javascript does in this case is irrelevant because of [this CSS hiding the element](https://github.com/guardian/frontend/blob/master/static/src/stylesheets/module/_adslot.scss#L443). It does not matter that it always gets inserted.

### Native ad with no advertisement label
screenshot_2018-08-02 how cinema embraced the silent agony of being a parent

### Normal MPU ad with missing advertisement label
screenshot_2018-08-02 how cinema embraced the silent agony of being a parent 1

### Tested

- [x] Locally
- [ ] On CODE (optional)

@guardian/commercial-dev

### CODE REVIEW

File: static/src/javascripts/projects/commercial/modules/dfp/Advert.js [hunk 1]
```diff
@@ -134,6 +137,34 @@ class Advert {
...
+        oldClasses: Array<string>,
+        newClasses: Array<string>
+    ): Array<string> {
+        const classesToRemove: Array<string> = [];
+        for (let i = 0; i < oldClasses.length; i += 1) {
+            const currentClass: string = oldClasses[i];
+            if (!newClasses.includes(currentClass)) {
+                classesToRemove.push(currentClass);
+            }
+        }
```
Comment (kelvin-chappell) [inline static/src/javascripts/projects/commercial/modules/dfp/Advert.js]: Can this be done with a filter?

Comment (janua) [reply static/src/javascripts/projects/commercial/modules/dfp/Advert.js]: Thanks! Stayed away with it as wasn't sure if polyfilled or not and initial implementation was more elaborate.

**Review by kelvin-chappell — APPROVED** [review]
👍

**Review by katebee — APPROVED** [review]
👍 Brilliant

Yay for more tests! ⭐️ 🚀

**Pull request merged**